### PR TITLE
Enable -Werror on s390x

### PR DESCRIPTION
**Pull request by AlekseiNikiforovIBM**
Enable -Werror on s390x

Example of original issue on s390x:
https://github.com/pytorch/pytorch/actions/runs/11014606340/job/30585632704

Most of warnings are not specific to s390x, but specific to gcc-13 or gcc-14. To test it on s390x an image with gcc-13 is needed. For s390x it's tested for new regressions on every merge due to trunk workflow.

`-Wdangling-reference` produces either obviously false warnings or suspicious warnings, which on closer inspection look plausibly safe.

`-Wredundant-move` with new gcc complains about `std::move(...)` disabling copy elision. But removing `std::move(...)` makes used clang versions complain about copying objects when they could be moved. For now also disable it.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @EikanWang

### CODE REVIEW

**Comment by AlekseiNikiforovIBM** [comment]
Fixes for different files are split into separate commits. Every commit has fixed errors listed in commit message.

**Review by malfet — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -1088,6 +1088,8 @@ if(NOT MSVC)
   append_cxx_flag_if_supported("-fno-math-errno" CMAKE_CXX_FLAGS)
   append_cxx_flag_if_supported("-fno-trapping-math" CMAKE_CXX_FLAGS)
   append_cxx_flag_if_supported("-Werror=format" CMAKE_CXX_FLAGS)
+  append_cxx_flag_if_supported("-Wno-error=dangling-reference" CMAKE_CXX_FLAGS)
+  append_cxx_flag_if_supported("-Wno-error=redundant-move" CMAKE_CXX_FLAGS)
```
Comment (malfet) [inline CMakeLists.txt:1092]: Let's restrict those to a specific specific GCC versions and if the are arechitecture specific let's restrict them to s360

**Review by malfet — CHANGES_REQUESTED** [review]
Please restrict warning suppressions in CMakeLists to a specific lists

File: torch/csrc/jit/serialization/export_module.cpp [hunk 2]
```diff
@@ -204,8 +204,8 @@ std::pair<IValue, IValue> getFunctionTuple(
       // For DictType, there are two items in t->containedTypes(), the first one
       // is key and the second one is value. Both of them could be NamedTuple
       // type.
-      const TypePtr& key_type = t->containedTypes()[0];
-      const TypePtr& value_type = t->containedTypes()[1];
+      const TypePtr key_type = t->containedTypes()[0];
```
[the remark this answers is not in the document]
Comment (AlekseiNikiforovIBM) [reply torch/csrc/jit/serialization/export_module.cpp]: I've removed this commit, replacing it with warning suppression. Is this commend still relevant?

**Comment by AlekseiNikiforovIBM** [comment]
@malfet @huydhn , could you please take a look at this PR again?

**Comment by malfet** [comment]
@AlekseiNikiforovIBM this PR reports 83 failures? Are you sure all of them are unrelated? Please give a reasonable explanaiton why or better rebase to `viable/strict`

**Review by malfet — APPROVED** [review]
4 changes files LGTM though

**Pull request closed without merging**